centreon/src/Core/ResourceAccess/Infrastructure/Repository/DbDatasetFilterFactory.php

```diff
@@ -0,0 +1,106 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For more information : contact@centreon.com
+ *
+ */
+
+declare(strict_types=1);
+
+namespace Core\ResourceAccess\Infrastructure\Repository;
+
+use Core\Infrastructure\Common\Repository\DbFactoryUtilitiesTrait;
+use Core\ResourceAccess\Domain\Model\DatasetFilter;
+
+/**
+ * @phpstan-import-type _DatasetFilter from DbReadRuleRepository
+ */
+class DbDatasetFilterFactory
+{
+    use DbFactoryUtilitiesTrait;
+
+    /**
+     * @param non-empty-array<_DatasetFilter> $record
+     *
+     * @throws \InvalidArgumentException
+     *
+     * @return DatasetFilter
+     */
+    public static function createFromRecord(array $record): DatasetFilter
+    {
+        $datasetFilter = null;
+
+        $buildDatasetFilterWithHierarchy = function (array $data, ?int $parentId, ?DatasetFilter $parentDatasetFilter) use (&$datasetFilter, &$buildDatasetFilterWithHierarchy): void {
+            if ($datasetFilter === null) {
+                /** @var non-empty-array<_DatasetFilter> $data */
+                $rootData = self::findRootFilter($data);
+                $datasetFilter = new DatasetFilter($rootData['dataset_filter_type'], self::fromStringToArrayOfInts($rootData['dataset_filter_resources']));
+                $buildDatasetFilterWithHierarchy($data, $rootData['dataset_filter_id'], $datasetFilter);
+            } else {
+                if (
+                    $parentId !== null
+                    && $parentDatasetFilter !== null
+                ) {
+                    /** @var non-empty-array<_DatasetFilter> $data */
+                    $childrenData = self::findSubFilter($parentId, $data);
+
+                    if ($childrenData !== []) {
+                        $childrenFilter = new DatasetFilter($childrenData['dataset_filter_type'], self::fromStringToArrayOfInts($childrenData['dataset_filter_resources']));
+                        $parentDatasetFilter->setDatasetFilter($childrenFilter);
+
+                        $buildDatasetFilterWithHierarchy($data, $childrenData['dataset_filter_id'], $parentDatasetFilter->getDatasetFilter());
+                    }
+                }
+            }
+        };
+
+        $buildDatasetFilterWithHierarchy($record, null, null);
+
+        /** @var DatasetFilter $datasetFilter */
+        return $datasetFilter;
+    }
+
+    /**
+     * @param int $parentId
+     * @param non-empty-array<_DatasetFilter> $data
+     *
+     * @return _DatasetFilter|array{}
+     */
+    private static function findSubFilter(int $parentId, array $data): array
+    {
+        $subFilter = array_values(
+            array_filter(
+                $data,
+                static fn (array $data): bool => $data['dataset_filter_parent_id'] === $parentId
+            )
+        );
+
+        return $subFilter !== [] ? $subFilter[0] : [];
```

Comment:
Other way to code that, seems cleaner to me. Up to you to do the change or not.
```suggestion
        return $subFilter[0] ?? [];
```